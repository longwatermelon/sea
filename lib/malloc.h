fn malloc(n:int) -> byte*;
fn free(p:byte*) -> void;
fn realloc(p:byte*, n:int) -> byte*;
fn memcpy(dst:byte*, src:byte*, n:int) -> void;
fn arena_reset() -> void;

