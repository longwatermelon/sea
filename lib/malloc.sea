#include "util.h"

let HEAP_CAP:int = 16777216;
let HEAP:byte* = galloc(16777216, byte);
let heap_top:int = 0;

let ALIGN:int = 8;
let HDR:int = 8;

fn align_up(x:int, a:int) -> int {
    let y:int = x + a - 1;
    return y - (y % a);
}

fn memcpy(dst:byte*, src:byte*, n:int) -> void {
    let i:int = 0;
    while (i<n) {
        dst[i] = src[i];
        i = i + 1;
    }
}

fn malloc(n:int) -> byte* {
    if (n<=0) { return 0; }

    let off:int = align_up(heap_top, ALIGN);
    let need:int = HDR + n;
    if (off + need > HEAP_CAP) { return 0; }

    let hdr:int* = HEAP + off;
    *hdr = n;

    heap_top = off + need;
    return (HEAP + off) + HDR;
}

fn free(p:byte*) -> void {
    if (p==0) { return 0; }

    let hdrp:byte* = p - HDR;
    let sz:int* = hdrp;
    let n:int = *sz;

    if ((p + n) == (HEAP + heap_top)) {
        heap_top = heap_top - (HDR + n);
    }
}

fn realloc(p:byte*, n:int) -> byte* {
    if (p==0) { return malloc(n); }
    if (n==0) { free(p); return 0; }

    let hdrp:byte* = p - HDR;
    let old_sz:int* = hdrp;
    let m:int = *old_sz;

    if ((p + m) == (HEAP + heap_top)) {
        let off:int = heap_top - (HDR + m);
        let need:int = HDR + n;
        if (off + need <= HEAP_CAP) {
            let hdr:int* = HEAP + off;
            *hdr = n;
            heap_top = off + need;
            return (HEAP + off) + HDR;
        }
    }

    let q:byte* = malloc(n);
    if (q==0) { return 0; }
    memcpy(q, p, min(m, n));
    free(p);
    return q;
}

fn arena_reset() -> void {
    heap_top = 0;
}
