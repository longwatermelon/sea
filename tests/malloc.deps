lib/malloc.sea
lib/util.sea
