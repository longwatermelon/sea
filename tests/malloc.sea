#include "../lib/malloc.h"

fn main() -> int {
    let too:byte* = malloc(16777216 + 1);
    if (too!=0) { return 1; }

    let a:int* = malloc(4*sizeof(int));
    a[0]=1; a[1]=2; a[2]=3; a[3]=4;
    let sum:int = a[0]+a[1]+a[2]+a[3];

    a = realloc(a, 8*sizeof(int));
    a[4]=5; a[5]=6; a[6]=7; a[7]=8;
    sum = sum + a[4]+a[5]+a[6]+a[7];

    free(a);
    return sum;
}

