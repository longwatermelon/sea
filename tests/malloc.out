RETURN_CODE:36
STDOUT:

